cli/src/cmd/cast/storage.rs

```diff
@@ -0,0 +1,206 @@
+    #[clap(help = "The contract address.", parse(try_from_str = parse_name_or_address), value_name = "ADDRESS")]
+    address: NameOrAddress,
+    #[clap(help = "The storage slot number (hex or decimal)", parse(try_from_str = parse_slot), value_name = "SLOT")]
+    slot: Option<H256>,
+    #[clap(long, env = "ETH_RPC_URL", value_name = "URL")]
+    rpc_url: Option<String>,
+    #[clap(
+        long,
+        short = 'B',
+        help = "The block height you want to query at.",
+        long_help = "The block height you want to query at. Can also be the tags earliest, latest, or pending.",
+        parse(try_from_str = parse_block_id),
+        value_name = "BLOCK"
+    )]
+    block: Option<BlockId>,
+
+    // Etherscan
+    #[clap(long, short, env = "ETHERSCAN_API_KEY", help = "etherscan API key", value_name = "KEY")]
+    etherscan_api_key: Option<String>,
+    #[clap(
+        long,
+        visible_alias = "chain-id",
+        env = "CHAIN",
+        help = "The chain ID the contract is deployed to.",
+        default_value = "mainnet",
+        value_name = "CHAIN"
+    )]
+    chain: Chain,
+
+    // Forge
+    #[clap(flatten)]
+    build: build::CoreBuildArgs,
+}
+
+impl StorageArgs {
+    pub async fn run(self) -> Result<()> {
+        let StorageArgs { address, block, build, rpc_url, slot, chain, etherscan_api_key } = self;
+
+        let rpc_url = consume_config_rpc_url(rpc_url);
+        let provider = try_get_http_provider(rpc_url)?;
+
+        let address = match address {
+            NameOrAddress::Name(name) => provider.resolve_name(&name).await?,
+            NameOrAddress::Address(address) => address,
+        };
+
+        // Slot was provided, perform a simple RPC call
+        if let Some(slot) = slot {
+            let cast = Cast::new(provider);
+            println!("{}", cast.storage(address, slot, block).await?);
+            return Ok(())
+        }
+
+        // No slot was provided
+
+        // Get deployed bytecode at given address
+        let address_code = provider.get_code(address, block).await?;
+        if address_code.is_empty() {
+            eyre::bail!("Provided address has no deployed code and thus no storage");
+        }
+
+        // Check if we're in a forge project
+        let project = build.project()?;
+        if project.paths.has_input_files() {
+            // Find in artifacts and pretty print
+            let project = with_storage_layout_output(project);
+            let out = compile(&project, false, false)?;
+            let artifact = out.artifacts().find(|(_, artifact)| match artifact.deployed_bytecode {
+                Some(ref deployed_code) => match deployed_code.bytecode {
+                    Some(ref bytecode) => match bytecode.object.as_bytes() {
+                        Some(bytes) => bytes == &address_code,
+                        None => false,
+                    },
+                    None => false,
+                },
+                None => false,
+            });
```

Comment:
this could be compressed simplified with an anonymous function here

```rust
fn is_match(artifact) -> Option<bool> {
   let bytes = artifact.deployed_bytecode.as_ref()?.bytecode.as_ref()?...
}

is_match(artifact).unwrap_or_default()
```

Reply:
nice thanks, didn't know you could do that

Reply:
also there should be a function in the `Artifact` trait that returns the code directly